lib/kadm5/iprop-log.8
.\" $Id$
.\"
.\" Copyright (c) 2005 - 2007 Kungliga Tekniska Högskolan
.\" (Royal Institute of Technology, Stockholm, Sweden).
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" 3. Neither the name of the Institute nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	$Id$
.\"
.Dd February 18, 2007
.Dt IPROP-LOG 8
.Os
.Sh NAME
.Nm iprop-log
.Nd examine and maintain the iprop log file
.Sh SYNOPSIS
.Nm
.Op Fl Fl version
.Op Fl h | Fl Fl help
.Ar command
.Pp
.Nm iprop-log truncate
.Oo Fl c Ar file \*(Ba Xo
.Fl Fl config-file= Ns Ar file
.Xc
.Oc
.Oo Fl r Ar string \*(Ba Xo
.Fl Fl realm= Ns Ar string
.Xc
.Oc
.Oo Fl K Ar integer \*(Ba Xo
.Fl Fl keep-entries= Ns Ar integer
.Xc
.Oc
.Oo Fl B Ar integer \*(Ba Xo
.Fl Fl max-bytes= Ns Ar integer
.Xc
.Oc
.Oo Fl R \*(Ba Xo
.Fl Fl reset
.Xc
.Oc
.Op Fl h | Fl Fl help
.Op Ar log-file
.Pp
.Nm iprop-log dump
.Oo Fl c Ar file \*(Ba Xo
.Fl Fl config-file= Ns Ar file
.Xc
.Oc
.Oo Fl r Ar string \*(Ba Xo
.Fl Fl realm= Ns Ar string
.Xc
.Oc
.Oo Fl n \*(Ba Xo
.Fl Fl no-lock
.Xc
.Oc
.Oo Fl R \*(Ba Xo
.Fl Fl reverse
.Xc
.Oc
.Oo Fl K Ar HOSTNAME\*(Ba Xo
.Fl Fl upstream=HOSTNAME
.Xc
.Oc
.Op Fl h | Fl Fl help
.Op Ar log-file
.Pp
.Nm iprop-log replay
.Op Fl Fl start-version= Ns Ar version-number
.Op Fl Fl end-version= Ns Ar version-number
.Oo Fl c Ar file \*(Ba Xo
.Fl Fl config-file= Ns Ar file
.Xc
.Oc
.Oo Fl r Ar string \*(Ba Xo
.Fl Fl realm= Ns Ar string
.Xc
.Oc
.Op Fl h | Fl Fl help
.Op Ar log-file
.Pp
.Nm iprop-log last-version
.Oo Fl c Ar file \*(Ba Xo
.Fl Fl config-file= Ns Ar file
.Xc
.Oc
.Oo Fl r Ar string \*(Ba Xo
.Fl Fl realm= Ns Ar string
.Xc
.Oc
.Oo Fl n \*(Ba Xo
.Fl Fl no-lock
.Xc
.Oc
.Op Fl h | Fl Fl help
.Op Ar log-files
.Pp
.Nm iprop-log signal
.Oo Fl c Ar file \*(Ba Xo
.Fl Fl config-file= Ns Ar file
.Xc
.Oc
.Oo Fl r Ar string \*(Ba Xo
.Fl Fl realm= Ns Ar string
.Xc
.Oc
.Oo Fl K Ar HOSTNAME\*(Ba Xo
.Fl Fl upstream=HOSTNAME
.Xc
.Oc
.Op Fl h | Fl Fl help
.Pp
.Sh DESCRIPTION
Supported options:
.Bl -tag -width Ds
.It Fl Fl version
.It Fl h , Fl Fl help
.El
.Pp
command can be one of the following:
.Bl -tag -width truncate
.It truncate
.Bl -tag -width Ds
.It Fl c Ar file , Fl Fl config-file= Ns Ar file
configuration file
.It Fl r Ar string , Fl Fl realm= Ns Ar string
realm
.It Xo
.Fl K Ar integer ,
.Fl Fl keep-entries= Ns Ar integer
.Xc
.It Xo
.Fl B Ar integer ,
.Fl Fl max-bytes= Ns Ar integer
.Xc
.It Xo
.Fl R ,
.Fl Fl reset
.Xc
.El
.Pp
If
.Fl Fl reset
is given, then the given, configured, or default log file will be
truncated and will start at version 1.  This forces full propagations to
all slave KDCs.
.Pp
Otherwise the log will be truncated but some entries will be preserved,
as specified by the
.Fl Fl keep-entries
and/or
.Fl Fl max-bytes
options.  The largest number of
.Fl Fl keep-entries
entries that are available and fit in the given
.Fl Fl max-bytes
option will be used.  The
.Fl Fl keep-entries option defaults to 100, and the
.Fl Fl max-bytes
option defaults to the
.Ar log-max-size
parameter in the configuration.
.Pp
.It dump
.Bl -tag -width Ds
.It Fl c Ar file , Fl Fl config-file= Ns Ar file
configuration file
.It Xo
.Fl r Ar string ,
.Fl Fl realm= Ns Ar string
.Xc
realm
.It Xo
.Fl n Ar string ,
.Fl Fl no-lock
.Xc
.It Xo
.Fl R Ar string ,
.Fl Fl reverse
.Xc
.It Xo
.Fl K Ar HOSTNAME ,
.Fl Fl upstream= Ns Ar HOSTNAME
.Xc
.El
.Pp
Print out all entries in the given, configured, or default log file to
standard output.  If the
.Fl n
option is used then don't lock the iprop log file.  If the
.Fl R
option is used, then print the entries in reverse order
(this can be useful when the log is very large).
.Pp
If the
.Fl K
option is given then the log that will be dumped is the one for
the given upstream (which normally exists only in multi-master
mode).
.It replay
.Bl -tag -width Ds
.It Fl Fl start-version= Ns Ar version-number
start replay with this version
.It Xo
.Fl Fl end-version= Ns Ar version-number
.Xc
end replay with this version
.It Fl c Ar file , Fl Fl config-file= Ns Ar file
configuration file
.It Fl r Ar string , Fl Fl realm= Ns Ar string
realm
.El
.Pp
Replay the changes from specified entries (or all if none is specified)
in the given, configured, or default transaction log file to the
database.
.It last-version
.Bl -tag -width Ds
.It Fl c Ar file , Fl Fl config-file= Ns Ar file
configuration file
.It Fl r Ar string , Fl Fl realm= Ns Ar string
realm
.It Xo
.Fl n Ar string ,
.Fl Fl no-lock
.Xc
.It Xo
.Fl K Ar HOSTNAME ,
.Fl Fl upstream= Ns Ar HOSTNAME
.Xc
.El
.Pp
prints the version of the last record in each of the given log files, or
the configured, or the default log file if none is given.
.Pp
If the
.Fl K
option is given then the log that will be show is the one for
the given upstream (which normally exists only in multi-master
mode).
.It signal
.Bl -tag -width Ds
.It Fl c Ar file , Fl Fl config-file= Ns Ar file
configuration file
.It Fl r Ar string , Fl Fl realm= Ns Ar string
realm
.El
.Pp
Signals the ipropd-master daemon to send updates to slaves.  Normally
kadmin does this every time it writes to the database, so this should
rarely be needed.
.El
.Sh SEE ALSO
.Xr iprop 8


NEWS
Release Notes - Heimdal - Version Heimdal 8.0

 Bug fixes

 - Errors found by the Coverity static analysis.
 - Errors found by the LLVM scan-build static analyzer.
 - Errors found by the valgrind memory debugger.
 - ASN.1 compiler support for IMPLICIT tagged named types.
 - Fix out-of-tree SQLite3 ccache permissions / umask issues.
 - Fix transit path validation CVE-2017-6594
 - iprop bugs, race conditions, and performance
 - Many misc. bugs

Features:

 - KDC: Add FAST support for TGS.
 - KDC: Greatly improved plugin facility for Samba.
 - KDC: Add httpkadmind service providing a subset of kadmin
        functionality over HTTP.
 - KDC: Add GSS-API pre-authentication support.
 - KDC: Add support for virtual service principal namespaces.
 - KDC: Add support for synthetic client principals that exist if the
        pre-authentication mechanism (e.g., PKINIT) can authenticate
        them, thus not requiring an HDB entry.
 - KDC: Between virtual service principal namespaces and synthetic client
        principals it is now possible to run a large realm with mostly empty
        HDBs containing only krbtgt principals -- no writes needed for users or
        services, and fast to re-key a realm in the event of compromise.
        In this mode, clients must use PKINIT or GSS-API pre-auth for this, and
        services must use PKINIT or GSS-API pre-auth to periodically fetch
        updated keytabs.
 - KDC: Revamp and enhance kx509 support (though bx509d mostly replaces kx509).
 - KDC: Better support for aliases and referrals.
 - KDC: Always return the salt in the PA-ETYPE-INFO[2].
 - KDC: Add warn_ticket_addresses configuration parameter.
 - KDC: allow anonymous AS requests with long-term keys.
 - KDC: Do not include PAC for anonymous AS requests.
 - KDC: Enable keepalive mode on incoming sockets.
 - KDC: Greatly improved logging.
 - KDC: Remove KRB5SignedPath, to be replaced with PAC.
 - PKIX: Add bx509d -- an online certification authority (CA) with an HTTP API.
 - kadmin: Add HTTP-based kadmin protocol.
 - kadmin: Add add_alias, del_alias.
 - kadmin: Add command aliases to man page.
 - kadmin: Add disallow-client attribute.
 - kadmin: add --hdb / -H argument.
 - kadmin: Allow enforcing password quality on admin password change.
 - kadmin: Improve ext_keytab usage.
 - kadmin: Selective pruning of historic key for principal.
 - krb5: Add client_aware_channel_bindings option.
 - krb5: Add constrained credential delegation option "destination TGT"
 - krb5: Add "EFILE:" target for logging.
 - krb5: Add include/includedir directives for krb5.conf.
 - krb5: Complete DIR ccache collection support.
 - krb5: Add FILE ccache collection support.
 - krb5: Improved FILE ccache performance.
 - krb5: Improve KCM ccache collection support.
 - krb5: Add KEYRING ccache support.
 - krb5: Add kx509 client.
 - krb5: Improve FILE keytab performance.
 - krb5: Implement KRB5_TRACE environment variable.
 - krb5: Add experimental name canonicalization rules configuration.
 - krb5: Support start_realm ccconfig entry type.
 - kinit: Add --default-for option for ccache collection support.
 - kinit: Add --pk-anon-fast-armor option.
 - kinit: Don't leave dangling temporary ccaches.
 - iprop: Many performance and scaling enhancements.
 - iprop: Support hierarchical propagation.
 - ASN.1: Document fuzzing process.
 - ASN.1: Complete template backend.
 - ASN.1: Add partial Information Object System support (template backend
          only).  This means that open type holes can be decoded recursively
          with one codec function call.

lib/kadm5/iprop-log.c
/*
 * Copyright (c) 1997 - 2005 Kungliga Tekniska Högskolan
 * (Royal Institute of Technology, Stockholm, Sweden).
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the Institute nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "iprop.h"
#include <sl.h>
#include <parse_time.h>
#include "iprop-commands.h"

RCSID("$Id$");

static krb5_context context;

static char *
get_iprop_log_for_upstream(const char *log_file, const char *upstream)
{
    const char *stem;
    size_t len;
    char *s = NULL;

    if (upstream == NULL) {
	if ((s = strdup(log_file)) == NULL)
	    krb5_errx(context, 1, "Out of memory");
        return s;
    }

    len = strlen(log_file);
    stem = strrchr(log_file, '.');
    if (stem == NULL || strcmp(stem, ".log") != 0)
        stem = log_file + len;
    if (asprintf(&s, "%.*s-%s%s", (int)(stem - log_file), log_file,
                 upstream, stem[0] == '.' ? stem : "") == -1 || s == NULL)
        krb5_errx(context, 1, "Out of memory");
    return s;
}

static kadm5_server_context *
get_kadmin_context(const char *config_file, char *realm)
{
    kadm5_config_params conf;
    krb5_error_code ret;
    void *kadm_handle;
    char *file = NULL;
    char **files;
    int aret;

    if (config_file == NULL) {
	aret = asprintf(&file, "%s/kdc.conf", hdb_db_dir(context));
	if (aret == -1 || file == NULL)
	    errx(1, "out of memory");
	config_file = file;
    }

    ret = krb5_prepend_config_files_default(config_file, &files);
    free(file);
    if (ret)
	krb5_err(context, 1, ret, "getting configuration files");

    ret = krb5_set_config_files(context, files);
    krb5_free_config_files(files);
    if (ret)
	krb5_err(context, 1, ret, "reading configuration files");

    memset(&conf, 0, sizeof(conf));
    if(realm) {
	conf.mask |= KADM5_CONFIG_REALM;
	conf.realm = realm;
    }

    ret = kadm5_init_with_password_ctx (context,
					KADM5_ADMIN_SERVICE,
					NULL,
					KADM5_ADMIN_SERVICE,
					&conf, 0, 0,
					&kadm_handle);
    if (ret)
	krb5_err (context, 1, ret, "kadm5_init_with_password_ctx");

    return (kadm5_server_context *)kadm_handle;
}

/*
 * dump log
 */

static const char *op_names[] = {
    "get",
    "delete",
    "create",
    "rename",
    "chpass",
    "modify",
    "randkey",
    "get_privs",
    "get_princs",
    "chpass_with_key",
    "nop"
};

static kadm5_ret_t
print_entry(kadm5_server_context *server_context,
	    uint32_t ver,
	    time_t timestamp,
	    enum kadm_ops op,
	    uint32_t len,
	    krb5_storage *sp,
	    void *ctx)
{
    char t[256];
    const char *entry_kind = ctx;
    int32_t mask;
    int32_t nop_time;
    uint32_t nop_ver;
    hdb_entry ent;
    krb5_principal source;
    char *name1, *name2;
    krb5_data data;
    krb5_context scontext = server_context->context;
    krb5_error_code ret;

    krb5_data_zero(&data);

    strftime(t, sizeof(t), "%Y-%m-%d %H:%M:%S", localtime(&timestamp));

    if((int)op < (int)kadm_get || (int)op > (int)kadm_nop) {
	printf("unknown op: %d\n", op);
	return 0;
    }

    printf ("%s%s: ver = %u, timestamp = %s, len = %u\n",
	    entry_kind, op_names[op], ver, t, len);
    switch(op) {
    case kadm_delete:
        ret = krb5_ret_principal(sp, &source);
        if (ret == 0)
            ret = krb5_unparse_name(scontext, source, &name1);
        if (ret)
            krb5_err(scontext, 1, ret, "Failed to read a delete record");
	printf("    %s\n", name1);
	free(name1);
	krb5_free_principal(scontext, source);
	break;
    case kadm_rename:
	ret = krb5_data_alloc(&data, len);
        if (ret == 0)
            ret = krb5_ret_principal(sp, &source);
        if (ret == 0 && krb5_storage_read(sp, data.data, data.length) == -1)
            ret = errno;
        if (ret == 0)
            ret = hdb_value2entry(scontext, &data, &ent);
        if (ret == 0)
            ret = krb5_unparse_name(scontext, source, &name1);
        if (ret == 0)
            ret = krb5_unparse_name(scontext, ent.principal, &name2);
        if (ret)
            krb5_err(scontext, 1, ret, "Failed to read a rename record");
	printf("    %s -> %s\n", name1, name2);
	free(name1);
	free(name2);
	krb5_free_principal(scontext, source);
	free_hdb_entry(&ent);
	break;
    case kadm_create:
	ret = krb5_data_alloc(&data, len);
        if (ret == 0 && krb5_storage_read(sp, data.data, data.length) == -1)
            ret = errno;
        if (ret == 0)
            ret = hdb_value2entry(scontext, &data, &ent);
	if (ret)
            krb5_err(scontext, 1, ret, "Failed to read a create record");
	mask = ~0;
	goto foo;
    case kadm_modify:
	ret = krb5_data_alloc(&data, len);
        if (ret == 0)
            ret = krb5_ret_int32(sp, &mask);
        if (ret == 0 && krb5_storage_read(sp, data.data, data.length) == -1)
            ret = errno;
        if (ret == 0)
            ret = hdb_value2entry(scontext, &data, &ent);
	if (ret)
            krb5_err(scontext, 1, ret, "Failed to read a modify record");
    foo:
	if(ent.principal /* mask & KADM5_PRINCIPAL */) {
	    ret = krb5_unparse_name(scontext, ent.principal, &name1);
            if (ret)
                krb5_err(scontext, 1, ret,
                         "Failed to process a create or modify record");
	    printf("    principal = %s\n", name1);
	    free(name1);
	}
	if(mask & KADM5_PRINC_EXPIRE_TIME) {
	    if(ent.valid_end == NULL) {
		strlcpy(t, "never", sizeof(t));
	    } else {
		strftime(t, sizeof(t), "%Y-%m-%d %H:%M:%S",
			 localtime(ent.valid_end));
	    }
	    printf("    expires = %s\n", t);
	}
	if(mask & KADM5_PW_EXPIRATION) {
	    if(ent.pw_end == NULL) {
		strlcpy(t, "never", sizeof(t));
	    } else {
		strftime(t, sizeof(t), "%Y-%m-%d %H:%M:%S",
			 localtime(ent.pw_end));
	    }
	    printf("    password exp = %s\n", t);
	}
	if(mask & KADM5_LAST_PWD_CHANGE) {
	}
	if(mask & KADM5_ATTRIBUTES) {
	    unparse_flags(HDBFlags2int(ent.flags),
			  asn1_HDBFlags_units(), t, sizeof(t));
	    printf("    attributes = %s\n", t);
	}
	if(mask & KADM5_MAX_LIFE) {
	    if(ent.max_life == NULL)
		strlcpy(t, "for ever", sizeof(t));
	    else
		unparse_time(*ent.max_life, t, sizeof(t));
	    printf("    max life = %s\n", t);
	}
	if(mask & KADM5_MAX_RLIFE) {
	    if(ent.max_renew == NULL)
		strlcpy(t, "for ever", sizeof(t));
	    else
		unparse_time(*ent.max_renew, t, sizeof(t));
	    printf("    max rlife = %s\n", t);
	}
	if(mask & KADM5_MOD_TIME) {
	    printf("    mod time\n");
	}
	if(mask & KADM5_MOD_NAME) {
	    printf("    mod name\n");
	}
	if(mask & KADM5_KVNO) {
	    printf("    kvno = %d\n", ent.kvno);
	}
	if(mask & KADM5_MKVNO) {
	    printf("    mkvno\n");
	}
	if(mask & KADM5_AUX_ATTRIBUTES) {
	    printf("    aux attributes\n");
	}
	if(mask & KADM5_POLICY) {
	    printf("    policy\n");
	}
	if(mask & KADM5_POLICY_CLR) {
	    printf("    mod time\n");
	}
	if(mask & KADM5_LAST_SUCCESS) {
	    printf("    last success\n");
	}
	if(mask & KADM5_LAST_FAILED) {
	    printf("    last failed\n");
	}
	if(mask & KADM5_FAIL_AUTH_COUNT) {
	    printf("    fail auth count\n");
	}
	if(mask & KADM5_KEY_DATA) {
	    printf("    key data\n");
	}
	if(mask & KADM5_TL_DATA) {
	    printf("    tl data\n");
	}
	free_hdb_entry(&ent);
	break;
    case kadm_nop :
        if (len == 16) {
            uint64_t off;
            ret = krb5_ret_uint64(sp, &off);
            if (ret)
                krb5_err(scontext, 1, ret, "Failed to read a no-op record");
            printf("uberblock offset %llu ", (unsigned long long)off);
        } else {
            printf("nop");
        }
        if (len == 16 || len == 8) {
            ret = krb5_ret_int32(sp, &nop_time);
            if (ret == 0)
                ret = krb5_ret_uint32(sp, &nop_ver);
            if (ret)
                krb5_err(scontext, 1, ret, "Failed to read a no-op record");

            timestamp = nop_time;
            strftime(t, sizeof(t), "%Y-%m-%d %H:%M:%S", localtime(&timestamp));
            printf("timestamp %s version %u", t, nop_ver);
        }
        printf("\n");
	break;
    default:
        krb5_errx(scontext, 1, "Unknown record type");
    }
    krb5_data_free(&data);

    return 0;
}

int
iprop_dump(struct dump_options *opt, int argc, char **argv)
{
    kadm5_server_context *server_context;
    krb5_error_code ret;
    enum kadm_iter_opts iter_opts_1st = 0;
    enum kadm_iter_opts iter_opts_2nd = 0;
    char *desc_1st = "";
    char *desc_2nd = "";

    server_context = get_kadmin_context(opt->config_file_string,
					opt->realm_string);

    {
	char *log_file;

	if (argc > 0)
	    log_file = strdup(argv[0]);
	else
	    log_file = strdup(server_context->log_context.log_file);
	if (log_file == NULL)
	    krb5_err(context, 1, errno, "Could not determine log file path");

	server_context->log_context.log_file =
	    get_iprop_log_for_upstream(log_file, opt->upstream_string);
	free(log_file);
    }
    if (server_context->log_context.log_file == NULL)
	krb5_err(context, 1, errno, "strdup");

    if (opt->reverse_flag) {
        iter_opts_1st = kadm_backward | kadm_unconfirmed;
        iter_opts_2nd = kadm_backward | kadm_confirmed;
        desc_1st = "unconfirmed ";
    } else {
        iter_opts_1st = kadm_forward | kadm_confirmed;
        iter_opts_2nd = kadm_forward | kadm_unconfirmed;
        desc_2nd = "unconfirmed";
    }

    if (opt->no_lock_flag) {
        ret = kadm5_log_init_sharedlock(server_context, LOCK_NB);
        if (ret == EAGAIN || ret == EWOULDBLOCK) {
            warnx("Not locking the iprop log");
            ret = kadm5_log_init_nolock(server_context);
            if (ret)
                krb5_err(context, 1, ret, "kadm5_log_init_nolock");
        }
    } else {
        warnx("If this command appears to block, try the --no-lock option");
        ret = kadm5_log_init_sharedlock(server_context, 0);
        if (ret)
            krb5_err(context, 1, ret, "kadm5_log_init_sharedlock");
    }

    ret = kadm5_log_foreach(server_context, iter_opts_1st,
                            NULL, print_entry, desc_1st);
    if (ret)
	krb5_warn(context, ret, "kadm5_log_foreach");

    ret = kadm5_log_foreach(server_context, iter_opts_2nd,
                            NULL, print_entry, desc_2nd);
    if (ret)
	krb5_warn(context, ret, "kadm5_log_foreach");

    ret = kadm5_log_end (server_context);
    if (ret)
	krb5_warn(context, ret, "kadm5_log_end");

    kadm5_destroy(server_context);
    return 0;
}

int
iprop_truncate(struct truncate_options *opt, int argc, char **argv)
{
    kadm5_server_context *server_context;
    krb5_error_code ret;
    time_t now;

    (void) krb5_timeofday(context, &now);
    server_context = get_kadmin_context(opt->config_file_string,
					opt->realm_string);

    if (argc > 0) {
        free(server_context->log_context.log_file);
        server_context->log_context.log_file = strdup(argv[0]);
        if (server_context->log_context.log_file == NULL)
            krb5_err(context, 1, errno, "strdup");
    }

    if (opt->keep_entries_integer < 0 &&
        opt->max_bytes_integer < 0) {
        opt->keep_entries_integer = 100;
        opt->max_bytes_integer = 0;
    }
    if (opt->keep_entries_integer < 0)
        opt->keep_entries_integer = 0;
    if (opt->max_bytes_integer < 0)
        opt->max_bytes_integer = 0;

    if (opt->reset_flag) {
        /* First recover unconfirmed records */
        ret = kadm5_log_init(server_context);
        if (ret == 0)
            ret = kadm5_log_reinit(server_context, 0, now);
    } else {
        ret = kadm5_log_init(server_context);
        if (ret)
            krb5_err(context, 1, ret, "kadm5_log_init");
        ret = kadm5_log_truncate(server_context, opt->keep_entries_integer,
                                 opt->max_bytes_integer);
    }
    if (ret)
	krb5_err(context, 1, ret, "kadm5_log_truncate");

    kadm5_log_signal_master(server_context);

    kadm5_destroy(server_context);
    return 0;
}

int
last_version(struct last_version_options *opt, int argc, char **argv)
{
    kadm5_server_context *server_context;
    char *alt_argv[2] = { NULL, NULL };
    krb5_error_code ret;
    uint32_t version, tstamp;
    size_t i;

    server_context = get_kadmin_context(opt->config_file_string,
					opt->realm_string);

    if (argc == 0) {
        alt_argv[0] = strdup(server_context->log_context.log_file);
        if (alt_argv[0] == NULL)
            krb5_err(context, 1, errno, "strdup");
        argv = alt_argv;
        argc = 1;
    }

    for (i = 0; i < argc; i++) {
	free(server_context->log_context.log_file);
	server_context->log_context.log_file =
	    get_iprop_log_for_upstream(argv[i], opt->upstream_string);
        if (server_context->log_context.log_file == NULL)
            krb5_err(context, 1, errno, "strdup");

        if (opt->no_lock_flag) {
            ret = kadm5_log_init_sharedlock(server_context, LOCK_NB);
            if (ret == EAGAIN || ret == EWOULDBLOCK) {
                warnx("Not locking the iprop log");
                ret = kadm5_log_init_nolock(server_context);
                if (ret)
                    krb5_err(context, 1, ret, "kadm5_log_init_nolock");
            }
        } else {
            warnx("If this command appears to block, try the "
                  "--no-lock option");
            ret = kadm5_log_init_sharedlock(server_context, 0);
            if (ret)
                krb5_err(context, 1, ret, "kadm5_log_init_sharedlock");
        }

        ret = kadm5_log_get_version (server_context, &version, &tstamp);
        if (ret)
            krb5_err (context, 1, ret, "kadm5_log_get_version");

        ret = kadm5_log_end (server_context);
        if (ret)
            krb5_warn(context, ret, "kadm5_log_end");

        printf("version: %lu (timestamp: %lu)\n", (unsigned long)version,
	       (unsigned long)tstamp);
    }

    kadm5_destroy(server_context);
    free(alt_argv[0]);
    return 0;
}

int
signal_master(struct signal_options *opt, int argc, char **argv)
{
    kadm5_server_context *server_context;

    server_context = get_kadmin_context(opt->config_file_string,
					opt->realm_string);

    kadm5_log_signal_master(server_context);

    kadm5_destroy(server_context);
    return 0;
}

/*
 * Replay log
 */

int start_version = -1;
int end_version = -1;

static kadm5_ret_t
apply_entry(kadm5_server_context *server_context,
	    uint32_t ver,
	    time_t timestamp,
	    enum kadm_ops op,
	    uint32_t len,
	    krb5_storage *sp,
	    void *ctx)
{
    struct replay_options *opt = ctx;
    krb5_error_code ret;

    if((opt->start_version_integer != -1 && ver < (uint32_t)opt->start_version_integer) ||
       (opt->end_version_integer != -1 && ver > (uint32_t)opt->end_version_integer)) {
	/* XXX skip this entry */
	return 0;
    }
    printf ("ver %u... ", ver);
    fflush (stdout);

    ret = kadm5_log_replay(server_context, op, ver, len, sp);
    if (ret)
	krb5_warn (server_context->context, ret, "kadm5_log_replay");

    printf ("done\n");

    return 0;
}

int
iprop_replay(struct replay_options *opt, int argc, char **argv)
{
    kadm5_server_context *server_context;
    krb5_error_code ret;

    server_context = get_kadmin_context(opt->config_file_string,
					opt->realm_string);

    if (argc > 0) {
        free(server_context->log_context.log_file);
        server_context->log_context.log_file = strdup(argv[0]);
        if (server_context->log_context.log_file == NULL)
            krb5_err(context, 1, errno, "strdup");
    }

    ret = server_context->db->hdb_open(context,
				       server_context->db,
				       O_RDWR | O_CREAT, 0600);
    if (ret)
	krb5_err (context, 1, ret, "db->open");

    ret = kadm5_log_init (server_context);
    if (ret)
	krb5_err (context, 1, ret, "kadm5_log_init");

    ret = kadm5_log_foreach(server_context,
                            kadm_forward | kadm_confirmed | kadm_unconfirmed,
                            NULL, apply_entry, opt);
    if(ret)
	krb5_warn(context, ret, "kadm5_log_foreach");
    ret = kadm5_log_end (server_context);
    if (ret)
	krb5_warn(context, ret, "kadm5_log_end");
    ret = server_context->db->hdb_close (context, server_context->db);
    if (ret)
	krb5_err (context, 1, ret, "db->close");

    kadm5_destroy(server_context);
    return 0;
}

static int help_flag;
static int version_flag;

static struct getargs args[] = {
    { "version", 	0,	arg_flag, 	&version_flag,
      NULL,		NULL
    },
    { "help", 	'h', 	arg_flag, 	&help_flag,
      NULL, NULL
    }
};

static int num_args = sizeof(args) / sizeof(args[0]);

int
help(void *opt, int argc, char **argv)
{
    if(argc == 0) {
	sl_help(commands, 1, argv - 1 /* XXX */);
    } else {
	SL_cmd *c = sl_match (commands, argv[0], 0);
 	if(c == NULL) {
	    fprintf (stderr, "No such command: %s. "
		     "Try \"help\" for a list of commands\n",
		     argv[0]);
	} else {
	    if(c->func) {
		static char shelp[] = "--help";
		char *fake[3];
		fake[0] = argv[0];
		fake[1] = shelp;
		fake[2] = NULL;
		(*c->func)(2, fake);
		fprintf(stderr, "\n");
	    }
	    if(c->help && *c->help)
		fprintf (stderr, "%s\n", c->help);
	    if((++c)->name && c->func == NULL) {
		int f = 0;
		fprintf (stderr, "Synonyms:");
		while (c->name && c->func == NULL) {
		    fprintf (stderr, "%s%s", f ? ", " : " ", (c++)->name);
		    f = 1;
		}
		fprintf (stderr, "\n");
	    }
	}
    }
    return 0;
}

static void
usage(int status)
{
    arg_printusage(args, num_args, NULL, "command");
    exit(status);
}

int
main(int argc, char **argv)
{
    int optidx = 0;
    krb5_error_code ret;

    setprogname(argv[0]);

    if(getarg(args, num_args, argc, argv, &optidx))
	usage(1);
    if(help_flag)
	usage(0);
    if(version_flag) {
	print_version(NULL);
	exit(0);
    }
    argc -= optidx;
    argv += optidx;
    if(argc == 0)
	usage(1);

    ret = krb5_init_context(&context);
    if (ret)
	errx(1, "krb5_init_context failed with: %d\n", ret);

    ret = sl_command(commands, argc, argv);
    if(ret == -1)
	warnx ("unrecognized command: %s", argv[0]);
    return ret;
}
